### PR TITLE
O1 style chain

### PR DESCRIPTION
**Pull request by KTS-o7**
In relation to #573 
Hello @willbakst 
I have written a rudimentary cookbook.  Currently right answers are a hit or miss. What am  I doing wrong ?

also when i followed Contribution.md i think `uv.lock` got updated. I have currently pushed it, should i be exlcuding it ?

sorry for this hurried notebook.

### CODE REVIEW

File: docs/tutorials/more_advanced/o1_style_thinking.ipynb [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from mirascope.core import groq\n",
+    "from datetime import datetime"
```
Comment (koxudaxi) [inline docs/tutorials/more_advanced/o1_style_thinking.ipynb:75]: Could you please import in the cell you will use?

File: docs/tutorials/more_advanced/o1_style_thinking.ipynb [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+    "def execute_task(steps: str) -> str:\n",
+    "    return f\"Execute these steps to solve the task. Wrap your thoughts in <think></think> tags: {steps}\"\n",
+    "\n",
+    "\n",
+    "@groq.call(\"llama-3.1-8b-instant\")\n",
+    "def final_answer(execution: str) -> str:\n",
+    "    return f\"Based on this execution, provide a final answer. Do not use <think></think> tags for the final answer: {execution}\"\n",
+    "\n",
+    "\n",
+    "def generate_cot_response(user_query):\n",
```
Comment (koxudaxi) [inline docs/tutorials/more_advanced/o1_style_thinking.ipynb:245]: Could you please addd type-hint for all functions?

**Review by koxudaxi — CHANGES_REQUESTED** [review]
@KTS-o7
Thank you for creatin the awsome PR.
I left comments for just a few things.

File: docs/tutorials/more_advanced/o1_style_thinking.ipynb (hunk 2 above)
Comment (KTS-o7) [reply docs/tutorials/more_advanced/o1_style_thinking.ipynb]: yes i have added type hints in the new commit

File: uv.lock [hunk 3]
```diff
@@ -4357,8 +4357,6 @@ version = "6.0.0"
 source = { registry = "https://pypi.org/simple" }
 sdist = { url = "https://files.pythonhosted.org/packages/18/c7/8c6872f7372eb6a6b2e4708b88419fb46b857f7a2e1892966b851cc79fc9/psutil-6.0.0.tar.gz", hash = "sha256:8faae4f310b6d969fa26ca0545338b21f73c6b15db7c4a8d934a5482faa818f2", size = 508067 }
 wheels = [
-    { url = "https://files.pythonhosted.org/packages/c5/66/78c9c3020f573c58101dc43a44f6855d01bbbd747e24da2f0c4491200ea3/psutil-6.0.0-cp27-none-win32.whl", hash = "sha256:02b69001f44cc73c1c5279d02b30a817e339ceb258ad75997325e0e6169d8b35", size = 249766 },
-    { url = "https://files.pythonhosted.org/packages/e1/3f/2403aa9558bea4d3854b0e5e567bc3dd8e9fbc1fc4453c0aa9aafeb75467/psutil-6.0.0-cp27-none-win_amd64.whl", hash = "sha256:21f1fb635deccd510f69f485b87433460a603919b45e2a324ad65b0cc74f8fb1", size = 253024 },
```
[the remark this answers is not in the document]
Comment (KTS-o7) [reply uv.lock:4361]: ok gotcha, i have added that too

**Comment by KTS-o7** [comment]
@koxudaxi I have made the changes you requested in the new commit.
Please let me know if any more changes are required

File: uv.lock [hunk 4]
```diff
@@ -784,9 +844,18 @@ wheels = [
...
+sdist = { url = "https://files.pythonhosted.org/packages/5f/51/913ecca3970a2227cf4d5e8937df52cc28f465ac442216110b8e3323262d/cloudpickle-2.2.1.tar.gz", hash = "sha256:d89684b8de9e34a2a43b3460fbca07d09d6e25ce858df4d5a44240403b6178f5", size = 60800 }
+wheels = [
+    { url = "https://files.pythonhosted.org/packages/15/80/44286939ca215e88fa827b2aeb6fa3fd2b4a7af322485c7170d6f9fd96e0/cloudpickle-2.2.1-py3-none-any.whl", hash = "sha256:61f594d1f4c295fa5cd9014ceb3a1fc4a70b0de1164b94fbc2d854ccba056f9f", size = 25944 },
+]
+
 [[package]]
 name = "cohere"
-version = "5.5.8"
+version = "5.11.0"
 source = { registry = "https://pypi.org/simple" }
```
Comment (koxudaxi) [inline uv.lock:859]: OK, I understand the reason of the CI error.
Recently, `cohere` sdk deleted the few class from definition.
But, mirascope uses the classes.

https://github.com/cohere-ai/cohere-python/pull/583/files#diff-1f6edbab9075bce38b9eb378de4aeb094ff60abf6e6d36b1e0d9fc11d1c029b3L167-R178

Comment (KTS-o7) [reply uv.lock:859]: So should  we plan on changing how mirascope interacts with Cohere ?

Comment (koxudaxi) [reply uv.lock]: I have created the PR to fix the package problem. 
https://github.com/Mirascope/mirascope/pull/620
And, I rebase the branch and change target branch of this PR.

File: mkdocs.yml [hunk 5]
```diff
@@ -175,6 +175,7 @@ nav:
           - Speech Transcription: "tutorials/more_advanced/speech_transcription.ipynb"
           - Query Plan: "tutorials/more_advanced/query_plan.ipynb"
           - Named Entity Recognition: "tutorials/more_advanced/named_entity_recognition.ipynb"
+          - o1 Style Thinking: "tutorials/more_advanced/o1_style_thinking.ipynb"
```
Comment (koxudaxi) [inline mkdocs.yml:178]: @KTS-o7 
Sorr, I forgot to mention one thing.

File: uv.lock (hunk 4 above)
Comment (KTS-o7) [reply uv.lock]: So lets consider this as resolved

File: mkdocs.yml (hunk 5 above)
Comment (KTS-o7) [reply mkdocs.yml:178]: Should i be doing anything 😅 Not so clear on what is expected of me now

**Comment by koxudaxi** [comment]
OK, we should wait for merged [support-cohere-version-5.11.0](https://github.com/Mirascope/mirascope/tree/support-cohere-version-5.11.0) branch to main.

**Comment by KTS-o7** [comment]
Yeah that is fine 
Also any chance of considering this repo for Hacktoberfest ?
If that's possible it would be really cool as it would help me reach the milestone.

File: mkdocs.yml (hunk 5 above)
Comment (koxudaxi) [reply mkdocs.yml:178]: I'm sorry for my comment is not clear.
We had to add the notebook in the `nav` section in `mkdocs.yml`.
After that, we can see the notebook in the menu.
image

**Comment by koxudaxi** [comment]
Thank you very much. Your contribution has been very helpful for the community.
I'm not very familiar with Hacktoberfest, so I'll ask @willbakst  about it.

File: mkdocs.yml (hunk 5 above)
Comment (KTS-o7) [reply mkdocs.yml:178]: Ok next time i will keep this in mind

File: docs/tutorials/more_advanced/o1_style_thinking.ipynb [hunk 6]
```diff
@@ -89,11 +89,11 @@
...
    "source": [
-    "from mirascope.core import groq\n",
     "from datetime import datetime\n",
-    "from typing import List, Dict\n",
     "\n",
-    "history: List[Dict[str, str]] = []\n",
+    "from mirascope.core import groq\n",
+    "\n",
+    "history: list[dict[str, str]] = []\n",
     "\n",
```
[the remark this answers is not in the document]
Comment (KTS-o7) [reply docs/tutorials/more_advanced/o1_style_thinking.ipynb:97]: Ok, marking it as resolved then

**Pull request closed without merging**